### PR TITLE
Keep artifacts of server(logs, snap, etc) if test has been failed

### LINKED ISSUE #197: Very hard to find files of the failed test after many passed tests
Instance folders in `var/nnn_<suite>/` are kept for all the running and finished tests. As a result, if a test hangs and there were 10 tests who passed before, each created a couple of instances, then you have > 20 folders like `master_nnnn` and `replica_xxxx`. Good luck with finding which pair is from the test which is hanging. They need to be grouped by tests somehow.

### LINKED ISSUE #253: luatest: do not remove test results on test fail
Some helpers (such as `cluster.drop()` ) should perform cleanup. It should be done only in case the test is passed. 
Otherwise all logs should be kept for the investigation.

### LINKED ISSUE #292: `server:stop()` won't work if the process is already dead
Example which I have in my private patch:
```Lua
    new_replica:start({wait_until_ready = false})
    local logfile = fio.pathjoin(new_replica.workdir,
                                 new_replica.alias .. '.log')
    t.helpers.retrying({}, function()
        assert(not new_replica.process:is_alive())
    end)
    t.assert(new_replica:grep_log(
        'Cluster name mismatch: expected test%-name, got <no%-name>', 1024,
        {filename = logfile}))
    -- Nullify already dead process or server:drop() fails.
    new_replica.process = nil
    new_replica:drop()
```
I have to nullify the `process` member or `:drop()` fails. This is because `:drop()` calls `Process.mt:kill()` which treats failure of `kill -s` command as an error. But it is not an error really (with how killing is implemented right now, at least).

### PR DESCRIPTION
**Pull request by ochaplashkin**
#### :heavy_minus_sign: 1. Not fail `server:drop()` if the process is already dead

Fixed the following error:

```lua
g.test_foo = function()
    s = Server:new({box_cfg = {bad_option}}
    s:start()  -- bad `box_cfg` value, the server process not started
    s:drop()  -- kill failed: 256
end
```

#### :heavy_minus_sign: 2. Remove the `server:clean()` method

The method was removed because there was an ambiguous understanding of the differences between the `stop()`, `drop()` and `clean()` methods.

Now, there are only two method:

- `drop()` - stops and perfoms a clean-up working directory of server object;
- `stop()` -  stops the server process and closes the connection.

#### :heavy_minus_sign:  3. Save server artifacts if test has been failed

If the test fails, all server artifacts (logs, xlogs, snapshots, etc.) will be saved to the `${VARDIR}/artifacts` directory for further analysis.

Default path for the test artifacts is `${VARDIR}/artifacts`.

How it will work:

- the test must contain any hook that calls the `server:drop()` method;
- if the test has been failed, server performs:
  - copy all server artifacts to the specific directory in the artifacts (`${VARDIR}/artifacts` by default);
  - if the copy was successful, delete the working directory of the server
     - else, raise an error.

```lua
-- test/foo_test.lua
g.before_test('foo', function()
    g.s1 = Server:new()
    g.s1:start()
    
    g.s2 = Server:new()
    g.s2:start()
    
    g.tom = Server:new({alias='tom'})
    g.tom:start()
end)

g.foo = function()
    t.assert_equals(true, false)
end

g.after_test('foo', function(g)
  g.s1:drop()
  -- your artifacts here: /tmp/t/artifacts/server-XXXX
  g.s2:drop()
  -- your artifacts here: /tmp/t/artifacts/server-YYYY
  g.tom:drop()
  -- your artifacts here: /tmp/t/artifacts/tom-AAAA
end)
```

####  :heavy_minus_sign:  4. Group server artifacts by replica set

Group the artifacts of replica set nodes for more convenient analysis after the tests fall.
By default, replica set artifacts will be located in the `${VARDIR}/artifacts/<replica-set/<nodes>` directory. 

####  :heavy_minus_sign:  `extra` Improve output printing for easier artifact search

Now, if you use server or replica set objects and your test fails, the console output will contain information about artifacts:

```
1) test_foo
stack traceback:
	...
	[C]: in function 'xpcall'
artifacts:
        <your_server_id> -> <your_path_to_server>
```

Just run:
```bash
$ ls <your_path_to_server>
00000000000000000000.snap  00000000000000000001.xlog
00000000000000000000.xlog  server.log
```

####  :heavy_minus_sign:  Server

See [test file](https://github.com/tarantool/luatest/blob/c9099bcd12128ebdc95fd64523f232bb854bcb27/test/collect_server_artifacts_test.lua#L17-L68)
```
1) collect_server_artifacts.test_foo
.../test/collect_server_artifacts_test.lua:42: expected: 2, actual: 1
stack traceback:
	...
	[C]: in function 'xpcall'
artifacts:
        all -> /tmp/t/artifacts/all-uKlsxAzscx5j
        all2 -> /tmp/t/artifacts/all2-g6cXoEkYUJL4
        each -> /tmp/t/artifacts/each-zRqrRmohkFro
        each2 -> /tmp/t/artifacts/each2-RvIefa3OIOu9
        test -> /tmp/t/artifacts/test-98d0AcfFoCJK
        test2 -> /tmp/t/artifacts/test2-la0zsKAeXsMX
```

####  :heavy_minus_sign:  Replica set

See [test file](https://github.com/tarantool/luatest/blob/c9099bcd12128ebdc95fd64523f232bb854bcb27/test/collect_rs_artifacts_test.lua)
```
1) collect_rs_artifacts.test_foo
..//test/collect_rs_artifacts_test.lua:91: expected: 2, actual: 1
stack traceback:
	...
	[C]: in function 'xpcall'
artifacts:
        replica1-all -> /tmp/t/artifacts/rs-nBfYcSkqz0HT/replica1-all-CKdeW4mKLJFl
        replica2-all -> /tmp/t/artifacts/rs-nBfYcSkqz0HT/replica2-all-ENptNk1o_Fpj
        replica3-all -> /tmp/t/artifacts/rs-nBfYcSkqz0HT/replica3-all-0po6gxYbb89j
        replica1-each -> /tmp/t/artifacts/rs-q8sxJKJuXZP3/replica1-each-gFcMfRMmdgC8
        replica2-each -> /tmp/t/artifacts/rs-q8sxJKJuXZP3/replica2-each-dS784hZBDtky
        replica3-each -> /tmp/t/artifacts/rs-q8sxJKJuXZP3/replica3-each-EWoIn5aSxVCT
        replica1-test -> /tmp/t/artifacts/rs-tLfVvFFuLYbK/replica1-test-2nL7XNt5fOGa
        replica2-test -> /tmp/t/artifacts/rs-tLfVvFFuLYbK/replica2-test-UN6eMW8Xb2NU
        replica3-test -> /tmp/t/artifacts/rs-tLfVvFFuLYbK/replica3-test-swDWCTzTN8gA
```

####  :heavy_minus_sign: Corner case with `before_all` hook

If the `before_all` hook fails, one of the tests from your test group will contain the output containing artifacts.

```lua
g.before_all(function()
    g.s = Server:new()
    g.s:start()
    t.assert_equals(1, 2) -- Boom!
end)

g.test_foo = function()
    t.assert_equals(1, 1) -- success
end 

g.test_bar = function()
    t.assert_equals(1, 1) -- success
end 

g.after_all(function()
    g.s:drop()
end)
```

Output:
```
Failed tests:
-------------

1) collect_server_artifacts.test_bar
//test/collect_server_artifacts_test.lua:18: expected: 2, actual: 1
stack traceback:
	...
	[C]: in function 'xpcall'
artifacts:
	all -> /tmp/t/artifacts/all-bqyAIZtJpOIh

2) collect_server_artifacts.test_foo
//test/collect_server_artifacts_test.lua:18: expected: 2, actual: 1
stack traceback:
	...
	[C]: in function 'xpcall'

```
#####  :heavy_minus_sign: For now we have non-deterministic behavior (#292) in the following case:

```lua
g.before_all(function()
    g.s = Server:new()
    g.s:start()
end)

g.test_bar = function()
    t.assert_equals(1, 2)
end

g.test_foo = function()
end

g.after_all(function()
    g.s:drop()
end)
```
If we are **lucky** with the sequence of running tests, then we will be able to see the artifacts.

- [x] Tests
- [x] Changelog
- [x] Documentation

Close #253
Close #197
Close #292

### CODE REVIEW

File: luatest/test_instance.lua [hunk 1]
```diff
@@ -16,6 +16,7 @@ end
 -- default constructor, test are PASS by default
 function TestInstance.mt:initialize()
     self.status = 'success'
+    self.artifacts = 'server artifacts:\n'
```
Comment (ylobankov) [inline luatest/test_instance.lua:19]: Shouldn't it be just `artifacts`?

File: luatest/utils.lua [hunk 2]
```diff
@@ -148,4 +149,10 @@ function utils.get_fn_location(fn)
     return ('%s:%s'):format(fn_source[#fn_source], fn_details.linedefined)
 end
 
+function utils.generate_id(length, urlsafe)
+    urlsafe = urlsafe or true
+    length = length or 9
```
Comment (ylobankov) [inline luatest/utils.lua:154]: Honestly, for me, it looks so-so. Let's use the following approach:

```lua
if not length then length = 9 end
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -49,6 +49,8 @@
 - Raise an error in the `Server:wait_for_condition()` function when
   the server process is terminated. This is useful to not wait for timeout, for example,
   when a server fails to start due to bad configuration.
+- Save server artifacts (logs, snapshots, etc.) if the test fails.
+- Group the working directory of servers into a replica set.
```
[the remark this answers is not in the document]
Comment (ochaplashkin) [reply CHANGELOG.md]: Yep, fixed.

File: luatest/output/junit.lua [hunk 4]
```diff
@@ -23,13 +23,17 @@ function Output.node_status_xml(node)
     if node:is('error') then
         return table.concat(
             {'            <error type="', Output.xml_escape(node.message), '">\n',
-             '                <![CDATA[', Output.xml_c_data_escape(node.trace),
-             ']]></error>\n'})
+            '                <![CDATA[', Output.xml_c_data_escape(node.trace),']]>\n',
```
[the remark this answers is not in the document]
Comment (ochaplashkin) [reply luatest/output/junit.lua]: Formatted it.

File: luatest/output/junit.lua [hunk 5]
```diff
@@ -23,13 +23,17 @@ function Output.node_status_xml(node)
...
+            '                <artifacts>', Output.xml_escape(node.artifacts),
```
[the remark this answers is not in the document]
Comment (ochaplashkin) [reply luatest/output/junit.lua]: After our discussion, there will be an empty tag here.

File: luatest/output/junit.lua [hunk 6]
```diff
@@ -23,13 +23,17 @@ function Output.node_status_xml(node)
...
+            '                </artifacts>\n',
+            '            </error>\n'})
     elseif node:is('fail') then
         return table.concat(
             {'            <failure type="', Output.xml_escape(node.message), '">\n',
-             '                <![CDATA[', Output.xml_c_data_escape(node.trace),
-             ']]></failure>\n'})
+             '                <![CDATA[', Output.xml_c_data_escape(node.trace),']]>\n',
```
[the remark this answers is not in the document]
Comment (ochaplashkin) [reply luatest/output/junit.lua]: Fixed too, thx!

File: luatest/output/tap.lua [hunk 7]
```diff
@@ -28,6 +28,7 @@ function Output.mt:update_status(node)
     end
     if (node:is('fail') or node:is('error')) and self.verbosity >= self.class.VERBOSITY.VERBOSE then
        print(prefix .. node.trace:gsub('\n', '\n' .. prefix))
+       print(prefix .. node.artifacts .. prefix)
```
[the remark this answers is not in the document]
Comment (ochaplashkin) [reply luatest/output/tap.lua]: Fixed.

File: luatest/output/text.lua [hunk 8]
```diff
@@ -60,6 +60,7 @@ function Output.mt:display_one_failed_test(index, fail) -- luacheck: no unused
     print(index..") " .. fail.name .. self.class.ERROR_COLOR_CODE)
     print(fail.message .. self.class.RESET_TERM)
     print(fail.trace)
+    print(fail.artifacts)
```
[the remark this answers is not in the document]
Comment (ochaplashkin) [reply luatest/output/text.lua]: Fixed too.

File: luatest/utils.lua (hunk 2 above)
Comment (ochaplashkin) [reply luatest/utils.lua]: Yep, fixed `length` and `urlsafe` params

File: luatest/utils.lua [hunk 9]
```diff
@@ -148,4 +149,10 @@ function utils.get_fn_location(fn)
     return ('%s:%s'):format(fn_source[#fn_source], fn_details.linedefined)
 end
 
+function utils.generate_id(length, urlsafe)
+    urlsafe = urlsafe or true
```
[the remark this answers is not in the document]
Comment (ochaplashkin) [reply luatest/utils.lua]: Fixed.

File: luatest/utils.lua [hunk 10]
```diff
@@ -1,5 +1,6 @@
 local fun = require('fun')
 local yaml = require('yaml')
+local digest = require('digest')
```
[the remark this answers is not in the document]
Comment (ochaplashkin) [reply luatest/utils.lua]: Moved it to the top.

File: luatest/server.lua [hunk 11]
```diff
@@ -362,6 +369,16 @@ end
 function Server:drop()
     self:stop()
 
+    local prefix = fio.pathjoin(Server.vardir, 'artifacts', self.rs_id or '')
+    if not self.current_test:is('success') then
+        local artifacts = fio.pathjoin(prefix, self.id)
+
+        local ok, err = fio.copytree(self.workdir, artifacts)
+        if not ok then
+            error(('Failed to copy directory: %s'):format(err))
```
[the remark this answers is not in the document]
Comment (ochaplashkin) [reply luatest/server.lua]: Formatted it.

File: luatest/server.lua [hunk 12]
```diff
@@ -183,6 +184,14 @@ function Server:initialize()
         -- so set it explicitly:
         self.env.LUATEST_LUACOV_ROOT = os.getenv('LUATEST_LUACOV_ROOT')
     end
+
+    if self.current_test == nil then
+        self.current_test = rawget(_G, 'current_test')
+
+        local prefix = fio.pathjoin(Server.vardir, 'artifacts', self.rs_id or '')
+        local artifacts = fio.pathjoin(prefix, self.id)
```
[the remark this answers is not in the document]
Comment (ochaplashkin) [reply luatest/server.lua]: Yep, fixed.

File: luatest/replica_set.lua [hunk 13]
```diff
@@ -4,9 +4,11 @@
 
 local checks = require('checks')
 local log = require('log')
+local fio = require('fio')
 
 local helpers = require('luatest.helpers')
 local Server = require('luatest.server')
+local utils = require('luatest.utils')
```
[the remark this answers is not in the document]
Comment (ochaplashkin) [reply luatest/replica_set.lua:11]: INHO, It is located where it is needed :)

File: luatest/server.lua [hunk 14]
```diff
@@ -104,7 +105,7 @@ end
 -- Initialize the server object.
 function Server:initialize()
     if self.id == nil then
-        self.id = digest.base64_encode(digest.urandom(9), {urlsafe = true})
+        self.id = ('%s-%s'):format(self.alias, utils.generate_id())
```
[the remark this answers is not in the document]
Comment (ochaplashkin) [reply luatest/server.lua]: Yep, my  misunderstanding it. Fixed.

File: luatest/server.lua [hunk 15]
```diff
@@ -275,7 +284,7 @@ end
 -- Optionally waits until the server is ready.
 --
 -- @tab[opt] params Parameters to restart the server with.
---   Like `command`, `args`, `env`, etc.
```
[the remark this answers is not in the document]
Comment (ochaplashkin) [reply luatest/server.lua]: Sorry, bad git diff :(

File: luatest/server.lua [hunk 16]
```diff
@@ -290,8 +299,6 @@ function Server:restart(params, opts)
         env = '?table',
         args = '?table',
         box_cfg = '?table',
-
-        http_port = '?number',
```
[the remark this answers is not in the document]
Comment (ochaplashkin) [reply luatest/server.lua]: Bad git diff too, sorry(

File: luatest/server.lua [hunk 17]
```diff
@@ -358,18 +358,13 @@ function Server:stop()
...
-    self.instance_id = nil
-    self.instance_uuid = nil
-end
-
 --- Stop the server and clean its working directory.
 function Server:drop()
     self:stop()
-    self:clean()
+
+    fio.rmtree(self.workdir)
```
Comment (ylobankov) [inline luatest/server.lua:383]: Let's have an empty line here.

File: luatest/server.lua [hunk 18]
```diff
@@ -183,6 +185,15 @@ function Server:initialize()
         -- so set it explicitly:
         self.env.LUATEST_LUACOV_ROOT = os.getenv('LUATEST_LUACOV_ROOT')
     end
+
+    if self.current_test == nil then
+        self.current_test = rawget(_G, 'current_test')
+        if self.current_test then
+            local prefix = fio.pathjoin(Server.vardir, 'artifacts', self.rs_id or '')
+            self.artifacts = fio.pathjoin(prefix, self.id)
+            self.current_test:add_server_artifacts_directory(self.alias, self.artifacts)
```
Comment (ylobankov) [inline luatest/server.lua:194]: `add_server_artifacts` will be enough.

File: luatest/utils.lua [hunk 19]
```diff
@@ -148,4 +149,10 @@ function utils.get_fn_location(fn)
     return ('%s:%s'):format(fn_source[#fn_source], fn_details.linedefined)
 end
 
+function utils.generate_id(length, urlsafe)
+    if urlsafe == nil then urlsafe = true end
```
Comment (ylobankov) [inline luatest/utils.lua:153]: Please swap this line with the line below.

Comment (ochaplashkin) [reply luatest/utils.lua]: Fixed.

File: luatest/server.lua (hunk 18 above)
Comment (ochaplashkin) [reply luatest/server.lua]: Fixed.

File: luatest/server.lua (hunk 17 above)
Comment (ochaplashkin) [reply luatest/server.lua:383]: Yep, formatted it.

**Review by ochaplashkin — COMMENTED** [review]
Ready to the next round

**Comment by ochaplashkin** [comment]
:warning: Due to the fact that we changed the external API of the server (removed the `clean()` method), we should not forget to fix the following tests in the tarantool:

- box-luatest
    - [gh_7800_recover_snap_without_spaces_test.lua](https://github.com/tarantool/tarantool/blob/master/test/box-luatest/gh_7800_recover_snap_without_spaces_test.lua#L13);
    - [gh_7846_crash_on_replica_id_alter_test.lua](https://github.com/tarantool/tarantool/blob/master/test/box-luatest/gh_7846_crash_on_replica_id_alter_test.lua#L49-L53);
    - [gh_4264_recursive_trigger_test.lua](https://github.com/tarantool/tarantool/blob/master/test/box-luatest/gh_4264_recursive_trigger_test.lua#L76);
    - [gh_7974_force_recovery_bugs_test.lua](https://github.com/tarantool/tarantool/blob/master/test/box-luatest/gh_7974_force_recovery_bugs_test.lua#L78);
    - [gh_6794_recover_nonmatching_xlogs_test.lua](https://github.com/tarantool/tarantool/blob/master/test/box-luatest/gh_6794_recover_nonmatching_xlogs_test.lua#L14);
    - [gh_7917_log_row_on_recovery_error_test.lua](https://github.com/tarantool/tarantool/blob/master/test/box-luatest/gh_7917_log_row_on_recovery_error_test.lua#L27);
- replication-luatest
    - [bootstrap_strategy_auto_test.lua](https://github.com/tarantool/tarantool/blob/master/test/replication-luatest/bootstrap_strategy_auto_test.lua#L117).

File: luatest/server.lua (hunk 17 above)
Comment (ylobankov) [reply luatest/server.lua:383]: It should be formatted in the relevant commit.

File: test/server_test.lua [hunk 20]
```diff
@@ -322,11 +322,9 @@ g.test_wait_when_server_is_not_running_by_bad_option = function()
     t.assert_equals(status, false)
     t.assert_str_contains(msg, expected_msg)
     t.assert_equals(s1.process:is_alive(), false)
-    s1:clean()
```
Comment (ylobankov) [inline test/server_test.lua:325]: Please put the `Not fail server:drop() if the process is already dead` commit before this one and change it to `:drop()`.

File: test/server_test.lua [hunk 21]
```diff
@@ -322,11 +322,9 @@ g.test_wait_when_server_is_not_running_by_bad_option = function()
...
 
     status, msg = pcall(Server.start, s2)
     t.assert_equals(status, false)
     t.assert_str_contains(msg, expected_msg)
     t.assert_equals(s2.process:is_alive(), false)
-    s2:clean()
```
Comment (ylobankov) [inline test/server_test.lua:331]: ditto

File: luatest/output/tap.lua [hunk 22]
```diff
@@ -24,10 +24,13 @@ function Output.mt:update_status(node)
...
     local prefix = '#   '
     if self.verbosity > self.class.VERBOSITY.QUIET then
-       print(prefix .. node.message:gsub('\n', '\n' .. prefix))
+        print(prefix .. node.message:gsub('\n', '\n' .. prefix))
     end
     if (node:is('fail') or node:is('error')) and self.verbosity >= self.class.VERBOSITY.VERBOSE then
-       print(prefix .. node.trace:gsub('\n', '\n' .. prefix))
+        print(prefix .. node.trace:gsub('\n', '\n' .. prefix))
+        if node.artifacts then
+            print(prefix .. node.artifacts)
```
Comment (ylobankov) [inline luatest/output/tap.lua:32]:
```lua
print(prefix .. node.artifacts:gsub('\n', '\n' .. prefix))
```
See my comment in test_instance.lua.

File: luatest/test_instance.lua [hunk 23]
```diff
@@ -24,6 +25,14 @@ function TestInstance.mt:update_status(status, message, trace)
     self.trace = trace
 end
 
+function TestInstance.mt:add_server_artifacts(alias, workdir)
+    local prepared_str = string.format('#\t%s -> %s\n', alias, workdir)
```
Comment (ylobankov) [inline luatest/test_instance.lua:29]:
```lua
    local srv_workdir = string.format('\n\t%s -> %s', alias, workdir)
    if not self.artifacts then
        self.artifacts = 'artifacts:'
    end
    self.artifacts = self.artifacts .. srv_workdir
```

Such an approach doesn't produce redundant `#` in text/tap/junit output.

File: test/server_test.lua (hunk 20 above)
Comment (ochaplashkin) [reply test/server_test.lua:325]: Commits have been swapped.

File: test/server_test.lua (hunk 21 above)
Comment (ochaplashkin) [reply test/server_test.lua:331]: fixed.

File: luatest/output/tap.lua (hunk 22 above)
Comment (ochaplashkin) [reply luatest/output/tap.lua]: Fixed.

File: luatest/server.lua (hunk 17 above)
Comment (ochaplashkin) [reply luatest/server.lua:383]: Moved it.

**Review by ylobankov — APPROVED** [review]
